File: GraphsJava/Kosaraju.java
class Solution
{
    //Function to find number of strongly connected components in the graph.
    public int kosaraju(int N, ArrayList<ArrayList<Integer>> adj)
    {
        //code here
        Stack<Integer> stack = new Stack<Integer>();
        boolean[] visited = new boolean[N];
        for (int i=0;i<N;i++){
            if (visited[i]==false){
                 dfs(adj,visited,i,stack);
            }
        }
        //to get the new graph with edges reversed to proceed to step 2
        ArrayList<ArrayList<Integer>> newGraph = transpose(N,adj);
        
        //Now applying dfs second time in specific order acc to stack and new graph
        visited = new boolean[N];
        int count = 0;
        while (stack.size()>0){
            int current = stack.pop();
            if (visited[current]==false){
                dfs(newGraph,visited,current);
                count++;
            }
        }
        return count;
    }
    
    
    //Step 1 filling stack based on end time : put in before backtracking
    public static void dfs(ArrayList<ArrayList<Integer>> graph, boolean[] visited, int curr, Stack<Integer> stack){
        visited[curr] = true;
        for (int neighbour:graph.get(curr)){
            if (visited[neighbour]==false){
                dfs(graph,visited,neighbour,stack);
            }
        }
        stack.push(curr);
    }
    
    public static void dfs(ArrayList<ArrayList<Integer>> graph, boolean[] visited, int curr){
        visited[curr] = true;
        for (int neighbour:graph.get(curr)){
            if (visited[neighbour]==false){
                dfs(graph,visited,neighbour);
            }
        }
    }
    
    public static ArrayList<ArrayList<Integer>> transpose(int N, ArrayList<ArrayList<Integer>> graph){
        ArrayList<ArrayList<Integer>> result = new ArrayList<>(); 
        for (int i=0;i<N;i++){
            result.add(new ArrayList<>());
        }
        for (int i=0;i<N;i++){
            for (int nbrs:graph.get(i)){
                result.get(nbrs).add(i);
            }
        }
        return result;
    }
    
}

File: C++/GraphsCPP/kosaraju.cpp
Kosaraju Algorithm to Find Strongly Connected Components - GFG

class Solution
{
	public:
	//Function to find number of strongly connected components in the graph.
	vector<int> order;
	
	void dfs(int src, vector<bool> &vis, vector<int> adj[]){
	    vis[src] = true;
	    
	    for(auto i: adj[src]){
	        if(!vis[i])
	            dfs(i, vis, adj);
	    }
	    
	    order.push_back(src);
	}
	
	void rdfs(int src, vector<bool> &visited, vector<int> adj[]){
	    visited[src] = true;
	    
	    for(auto i: adj[src]){
	        if(!visited[i])
	            dfs(i, visited, adj);
	    }
	}
	
    int kosaraju(int V, vector<int> adj[]){
        //code here
        order.clear();
        
        vector<int> revAdj[V];
        for(int i = 0; i < V; i++){
            for(auto x: adj[i]){
                revAdj[x].push_back(i);
            }
        }

        vector<bool> vis(V, false);
        for(int i = 0; i < V; i++){
            if(!vis[i]){
                dfs(i, vis, adj);
            }
        }
        
        vector<bool> rvis(V, false);
        int count = 0;
        for(int i = order.size()-1; i >= 0; i--){
            if(!rvis[order[i]]){
                rdfs(order[i], rvis, revAdj);
                count++;
            }
        }
        
        return count;
    }
};
